internal/util/http_test.go

```diff
@@ -0,0 +1,234 @@
+package util
+
+/*
+ Copyright 2021 Crunchy Data Solutions, Inc.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/wojas/genericr"
+	"gotest.tools/v3/assert"
+
+	"github.com/crunchydata/postgres-operator/internal/logging"
+)
+
+func init() {
+	client = &MockClient{Timeout: 1}
+}
+
+func TestAddHeader(t *testing.T) {
+	t.Run("successful", func(t *testing.T) {
+		req := &http.Request{
+			Header: http.Header{},
+		}
+		versionString := "1.2.3"
+
+		result, _ := addHeader(req, versionString)
+		assert.DeepEqual(t,
+			result.Header[clientHeader][0],
+			`{"version":"1.2.3"}`,
+		)
+	})
+}
+
+func TestGetDurationFromEnvOrConstant(t *testing.T) {
+	orig := os.Getenv(updateCheckPeriodEnvVar)
+	os.Unsetenv(updateCheckPeriodEnvVar)
+	log := logging.Discard()
+
+	// Default
+	assert.Equal(t, getDurationFromEnvOrConstant(log,
+		updateCheckPeriodEnvVar, updateCheckPeriod),
+		24*time.Hour)
+
+	// Parseable period
+	os.Setenv(updateCheckPeriodEnvVar, "12h")
+	assert.Equal(t, getDurationFromEnvOrConstant(log,
+		updateCheckPeriodEnvVar, updateCheckPeriod),
+		12*time.Hour)
+
+	// Unparseable period
+	os.Setenv(updateCheckPeriodEnvVar, "argle")
+	assert.Equal(t, getDurationFromEnvOrConstant(log,
+		updateCheckPeriodEnvVar, updateCheckPeriod),
+		24*time.Hour)
+
+	os.Setenv(updateCheckPeriodEnvVar, orig)
+}
+
+func TestGetBackoff(t *testing.T) {
+	t.Run("default", func(t *testing.T) {
+		log := logging.Discard()
+
+		result := getBackoff(log)
+		assert.DeepEqual(t, result.Duration, 1*time.Second)
+		assert.DeepEqual(t, result.Steps, 4)
+		assert.DeepEqual(t, result.Cap, 30*time.Second)
+	})
+}
+
+type MockClient struct {
+	Timeout time.Duration
+}
+
+var funcFoo func() (*http.Response, error)
+
+// Do is the mock request that will return a mock success
+// TODO(benjaminjb): return to this when Operator Version API is complete
+func (m *MockClient) Do(req *http.Request) (*http.Response, error) {
+	return funcFoo()
+}
+
+func TestCheckForUpdates(t *testing.T) {
+	log := logging.Discard()
+	// Setting retries to 2 to speed up testing
+	orig := os.Getenv(updateCheckMaxRetriesEnvVar)
+	os.Setenv(updateCheckMaxRetriesEnvVar, "2")
+	backoff := getBackoff(log)
```

Comment:
🤔 This gets backoff parameters from the environment. Will the test behave differently in different environments?

Normally, the environment for a running process won't change. Perhaps we can avoid the above question by keeping settings in a variable/instance rather than loading them every call? Like a type that has the backoff settings, maybe a client, and methods.

```go
func (checker) CheckForUpdates()
type checker struct {
  backoff wait.Backoff
  client  *http.Client

  url, version string
}
```

The tests here can assign any backoff, client, etc, and not worry about restoring env vars. A "new-from-env" constructor could do all the parsing and keep behavior similar to now: `New(…).CheckForUpdates()`.

Reply:
This comes back to the issues I was wrestling with last week: on one hand, isn't it great to have things that are configurable but also provide defaults so you don't have to configure them; but on the other, under what circumstance would anyone really want to change these settings (besides the core configuration of turning this on/off)? 

We might want different settings for testing, but an even easier way to handle that is to just make them vars, set them one way in the code and another way in the testing, which again avoids any question of the environment accidentally setting these one way or another.